### PR TITLE
fix [`mem_replace_option_with_none`] not considering field variables

### PR DESCRIPTION
**Pull request by J-ZhengLi**
fixes: #9824 

---

changelog: fix [`mem_replace_option_with_none`] not considering field variables

### CODE REVIEW

File: clippy_lints/src/mem_replace.rs [hunk 1]
```diff
@@ -101,40 +101,26 @@ declare_clippy_lint! {
...
+    let mut applicability = Applicability::MachineApplicable;
+    span_lint_and_sugg(
+        cx,
+        MEM_REPLACE_OPTION_WITH_NONE,
+        expr_span,
+        "replacing an `Option` with `None`",
+        "consider `Option::take()` instead",
+        format!(
+            "{}.take()",
+            snippet_with_applicability(cx, sugg_expr_span, "", &mut applicability)
```
Comment (Jarcho) [inline clippy_lints/src/mem_replace.rs:120]: Should be `Sugg::hir_with_context(peel_ref_operators(cx, dest), expr_span.ctxt(), "", &mut applicability).mabe_par()` to handle precedence issues and macro calls.

Comment (J-ZhengLi) [reply clippy_lints/src/mem_replace.rs:120]: Thanks~ I do like this better, but I don't know the difference between them and in what case that parenthesis are needed, could you maybe provide one example? Which I'll probably add as one of the test cases.

Comment (Jarcho) [reply clippy_lints/src/mem_replace.rs:120]: `replace(&mut *x, None)` is the only practical example. If `x` is a type implementing `Deref` then the deref can't always be stripped out. This leaves you with `(*x).take()` as the required suggestion.

`x + x` could also return `&mut Option<_>`, but that's neither common nor recommended.

Comment (J-ZhengLi) [reply clippy_lints/src/mem_replace.rs]: Ahh, that's interesting, I completely forgot that's the case.
also, I've add a related test case

**Comment by Jarcho** [comment]
Looks good. Thank you. @bors r+

**Pull request merged**